### PR TITLE
Update NativeArray for toReversed()

### PR DESCRIPTION
**Pull request by pkkht**
Hi, 
I get the below error when testing the method manually from the console. I don't see NativeArray invoked in the exception stack trace. What's going wrong in the code? Any ideas please

js> const items = [1, 2, 3];
js> items
1,2,3
js> items.toReversed()
Exception in thread "main" java.lang.ArrayIndexOutOfBoundsException: Index 70 out of bounds for length 70
        at org.mozilla.javascript.IdScriptableObject$PrototypeValues.ensureId(IdScriptableObject.java:284)
        at org.mozilla.javascript.IdScriptableObject$PrototypeValues.get(IdScriptableObject.java:162)
        at org.mozilla.javascript.IdScriptableObject.get(IdScriptableObject.java:380)
        at org.mozilla.javascript.ScriptableObject.getProperty(ScriptableObject.java:2037)
        at org.mozilla.javascript.ScriptRuntime.getPropFunctionAndThisHelper(ScriptRuntime.java:2584)
        at org.mozilla.javascript.ScriptRuntime.getPropFunctionAndThis(ScriptRuntime.java:2575)
        at org.mozilla.javascript.optimizer.OptRuntime.callProp0(OptRuntime.java:71)
        at org.mozilla.javascript.gen._stdin__3._c_script_0(<stdin>:4)
        at org.mozilla.javascript.gen._stdin__3.call(<stdin>)
        at org.mozilla.javascript.ContextFactory.doTopCall(ContextFactory.java:383)
        at org.mozilla.javascript.ScriptRuntime.doTopCall(ScriptRuntime.java:3876)
        at org.mozilla.javascript.gen._stdin__3.call(<stdin>)
        at org.mozilla.javascript.gen._stdin__3.exec(<stdin>)
        at org.mozilla.javascript.tools.shell.Main.processSource(Main.java:482)
        at org.mozilla.javascript.tools.shell.Main.processFiles(Main.java:180)
        at org.mozilla.javascript.tools.shell.Main$IProxy.run(Main.java:98)
        at org.mozilla.javascript.Context.call(Context.java:544)
        at org.mozilla.javascript.ContextFactory.call(ContextFactory.java:475)
        at org.mozilla.javascript.tools.shell.Main.exec(Main.java:164)
        at org.mozilla.javascript.tools.shell.Main.main(Main.java:142)

### CODE REVIEW

**Comment by p-bakker** [comment]
I'd set a breakpoint early in the stacktrace you're getting and then see what is the difference between a call to [1,2,3].reverse() and [1,2,3].toReversed().

If you do, I think you'll find that you forgot to update MAX_PROTOTYPE_ID to the new value

**Comment by pkkht** [comment]
> I'd set a breakpoint early in the stacktrace you're getting and then see what is the difference between a call to [1,2,3].reverse() and [1,2,3].toReversed().
> 
> If you do, I think you'll find that you forgot to update MAX_PROTOTYPE_ID to the new value

cool, thanks. I know it is a basic question, but how to debug from the shell? Is it this one:

https://rhino.github.io/tools/debugger/

Says there is no breakpoint available?

**Comment by p-bakker** [comment]
That's the JavaScript Debugger. In your case as you're developing Rhino itself, in Java, you'll need a Java debugger.

I personally use Eclipse and run the rhino shell from source inside Eclipse, but you can use whatever Java development environment you like

File: src/org/mozilla/javascript/NativeArray.java [hunk 1]
```diff
@@ -2252,6 +2261,18 @@ private static boolean js_isArray(Object o) {
         return "Array".equals(((Scriptable) o).getClassName());
     }
 
+    private static Scriptable js_toReversed(
+            Context cx, Scriptable scope, Scriptable thisObj) {
+        Scriptable o = ScriptRuntime.toObject(cx, scope, thisObj);
+        long len = (getLengthProperty(cx, o) > Integer.MAX_VALUE) ? 0 : getLengthProperty(cx, o);
```
Comment (rbri) [inline src/org/mozilla/javascript/NativeArray.java:2267]: i guess getLengthProperty(cx, o) is expensive, maybe calling this only once is a good idea

**Review by rbri — COMMENTED** [review]
Test cases are missing

File: src/org/mozilla/javascript/NativeArray.java [hunk 2]
```diff
@@ -2729,7 +2753,8 @@ protected int findPrototypeId(String s) {
             Id_at = 32,
             Id_flat = 33,
             Id_flatMap = 34,
-            SymbolId_iterator = 35,
+            Id_toReversed =36,
```
Comment (p-bakker) [inline src/org/mozilla/javascript/NativeArray.java:2756]: formatting

File: src/org/mozilla/javascript/NativeArray.java [hunk 3]
```diff
@@ -2729,7 +2753,8 @@ protected int findPrototypeId(String s) {
             Id_at = 32,
             Id_flat = 33,
             Id_flatMap = 34,
-            SymbolId_iterator = 35,
```
Comment (p-bakker) [inline src/org/mozilla/javascript/NativeArray.java:2732]: Leave this, just update MAX_PROTOTYPE_ID to be Id_toReversed

**Comment by pkkht** [comment]
> Test cases are missing

Hi there, I am looking the array-at.js file. Probably, a similar file for this new method? How do I run just the test js file?

**Comment by rbri** [comment]
@pkkht you can have a look at the packages org.mozilla.javascript.tests.es2019, org.mozilla.javascript.tests.es2020, org.mozilla.javascript.tests.es2022. Check out e.g. org.mozilla.javascript.tests.es2020.TypedArrayAtTest for a sample.

**Comment by p-bakker** [comment]
@pkkht I see you mentioned this PR being a draft, but it's not marked as such and it's referencing just the .toReversed() method

What is your idea? Are you going to implement all the new methods of the Change Array by Copy proposal on NativeArray in this PR? Or are you planning to provide individual PRs for each method?

**Comment by p-bakker** [comment]
As for the tests: the test262 testsuite for [Array.prototype.toReversed()](https://github.com/tc39/test262/tree/main/test/built-ins/Array/prototype/toReversed) is (obviously) way more extensive than the tests you've added.

This is not particularly an issue with the tests you've added, but it does make me wonder what we might be missing. One thing in particular with our NativeArray implementation is that it internally switches to another implementation if the length of an Array is over NativeArray.maximumInitialCapacity (also see the NativeArray.denseOnly flag. I wonder if we should have explicit tests with Arrays that go beyond this size

I will see if I can expedite the work on #958 which hopefully would allow us to move to the latest or at least way more recent version of Test262

**Comment by pkkht** [comment]
Ah, I thought of adding other methods in the next PR. Is that ok

**Comment by pkkht** [comment]
Also did the workflow pass?

**Comment by pkkht** [comment]
Okay, on a second thought, I will use the same PR for all the methods. I see there is more work in the tests.. so we can do altogether

**Comment by pkkht** [comment]
Why did the tests fail in the pipeline?

File: src/org/mozilla/javascript/NativeArray.java [hunk 4]
```diff
@@ -2729,8 +2752,9 @@ protected int findPrototypeId(String s) {
...
-            MAX_PROTOTYPE_ID = SymbolId_iterator;
+            Id_toReversed = 36,
+            SymbolId_iterator = 37,
+            MAX_PROTOTYPE_ID = Id_toReversed;
     private static final int ConstructorId_join = -Id_join,
```
Comment (rbri) [inline src/org/mozilla/javascript/NativeArray.java:2758]: Id_flatMap = 34,
            Id_toReversed = 35,
            SymbolId_iterator = 36,
            MAX_PROTOTYPE_ID = SymbolId_iterator;

Comment (pkkht) [reply src/org/mozilla/javascript/NativeArray.java:2758]: Ta

**Comment by p-bakker** [comment]
@pkkht hi, wondering where you're at with this PR. Are you still working on it?

**Comment by pkkht** [comment]
Not as of now

**Comment by p-bakker** [comment]
@pkkht note that Rhino now runs the Test262 test suite against a very recent version of test262, which includes all the tests for newer Array methods like .at() and .toReversed()

**Comment by pkkht** [comment]
Okay

**Comment by rbri** [comment]
@pkkht why did you close this? Looks like the implementation of toReversed is still missing in Rhino and i think your PR will at least help to get this done.

**Comment by pkkht** [comment]
Sorry I got a bit confused with the recent comments.
Regardless, I am unable to take forward the code change at this stage.

**Comment by andreabergia** [comment]
@rbri I am working on a PR for this (and the other new `toXXX` methods: `toSorted`, `toSpliced`, `with`) by the way. I plan to open it this week, unless there's some emergency at work. 🙂

**Pull request closed without merging**